ROX-20835: ADR: update CVSS data updating pipeline to adpot NVD API (#8065)
stackrox/stackrox: scanner/decisions/0005-scannerv4-read-cvss-data.md

# 0005 - Adapt Scanner V4 NVD CVSS data updating pipeline to integrate with NVD CVE Api

- **Author(s):** Yi Li <yli3@redhat.com>
- **Created:** [2023-10-10]

## Context

The National Vulnerability Database (NVD) has announced the retirement of its JSON feeds by December 15, 2023. This creates the need to change our existing CVSS data updater pipeline for Scanner V4, which currently relies on these JSON feeds.

## Decision

The NVD CVSS Updater will retrieve CVSS data from Google storage at a configurable interval. It will make a single HTTP call for each NVD data bundle categorized by CVSS V3 severity. As an illustration, all CVE data with a CVSS V3 severity of 'low' will be archived into 'severity-low.zip' in Google storage and then being fetched by the updater.

> A GitHub workflow will keep the NVD data bundle in the Google Storage up-to-date. This Github workflow send http request to NVD CVE api to get data based on each level of CVSS V3 severities. Such as "curl 'https://services.nvd.nist.gov/rest/json/cves/2.0?cvssV3Severity=CRITICAL&startIndex=0' > severity-critical-0.json".  Per the NVD's documentation, the NVD CVE API omits CVSS v3 vector strings with a 'NONE' severity. This ensures we only obtain data bearing valid CVSS V3 metrics. For JSON files of a singular severity, we compress them into a single zip file, such as severity-low.zip.

Reviewer: > Per the NVD's documentation

Nit: It would be nice to have docs linked here.

> This updater will download zip files corresponding to four distinct severity levels. Then it will write each json file from these zipped archives to a single zip file in Central's file system. We will not need to do any json parsing in the updater. This compressed file is typically around 50 MB in size.

Reviewer: Just to confirm:

> This updater will download zip files corresponding to four distinct severity levels.

"Low", "Medium", "High", and "None"? I understood that the API filters scores with `severity == "None"`:

 > Per the NVD's documentation, the NVD CVE API omits CVSS v3 vector strings with a 'NONE' severity.

Author: "Low", "Medium", "High", and "Critical"

## Consequences
Transitioning from the NVD JSON feed to the NVD CVE API introduces certain constraints in data retrieval within the GitHub Action workflow, such as rate limiting and pagination. As a result, we will encounter a higher number of smaller, segmented JSON files.

> Changing from the NVD JSON feed to the NVD CVE API also give us the conveniences of no need to filter out data with out valid CVSS V3 metrics. 

Reviewer: Nit: Space above this line.

# 0005 - Adapt Scanner V4 NVD CVSS data updating pipeline to integrate with NVD CVE Api

- **Author(s):** Yi Li <yli3@redhat.com>
- **Created:** [2023-10-10]

## Status

Accepted

## Context

The National Vulnerability Database (NVD) has announced the retirement of its JSON feeds by December 15, 2023. This creates the need to change our existing CVSS data updater pipeline for Scanner V4, which currently relies on these JSON feeds.

## Decision

> The current NVD CVSS GitHub Workflow will be updated to fetch data from the NVD CVE API instead of NVD Json feeds while maintaining the freshness of the NVD CVSS data bundle in Google Storage. It will categorize the downloaded data by CVSS V3 severity levels, and these categorized data will be archived into individual zip files, such as `severity-low.zip`. The format of the JSON is the same as before.

Reviewer: > The current NVD CVSS GitHub Workflow will be updated to fetch data from the NVD CVE API instead of NVD Json feeds while maintaining the freshness of the NVD CVSS data bundle in Google Storage.

Please briefly explain the diff between the JSON endpoints and the NVD API. With that information, I could give you feedback on the decision to create individual ZIPs according to CVSS V3 severity levels. I am curious to understand why grouping by severity. And also, if there is any implication on the reliability of the workflows to handle API issues that might block the GitHub workflow from updating the bucket (e.g., rate limiting).

> The CVSS Updater in Central will be modified to download zip files corresponding to four distinct severity levels and write them into a single zip file stored in Central's file system. This eliminates the need for JSON parsing in the updater. ScannerV4 will pull the zip bundle and populate the Matcher DB with the enrichment data.

Reviewer: > The CVSS Updater in Central will be modified to download zip files corresponding to four distinct severity levels and write them into a single zip file stored in Central's file system.

In other words, you're making one  HTTP request for each `<severity>.zip`. What happens if one of these GETs fails when generating the single zip?

Author: We need to make multiple requests for each severity.

Reviewer: > We need to make multiple requests for each severity.

Let's break it down.

Success scenario:

| What | Result | What to do |
|-|-|-|
| GET /bucket/<first>-severity.zip | Success | Keep in memory to add to the final zip file |
| GET /bucket/<second>-severity.zip | Success | Keep in memory to add to the final zip file |
| GET /bucket/<third>-severity.zip | Success |  Keep in memory to add to the final zip file |
| Create a final zip file from all zips in memory | Success | Write zip file to the filesystem and serve upon request from Scanner. |

Scenario I want to discuss:

| What | Result | What to do |
|-|-|-|
| GET /bucket/<first>-severity.zip | Success | Keep in memory to add to the final zip file |
| GET /bucket/<second>-severity.zip | Success | Keep in memory to add to the final zip file |
| GET /bucket/<third>-severity.zip | Failed (e.g. 401, 403, 505, Timeout) |  A |
| Create a final zip file from all zips in memory | B | C |

Could you please fill A, B and C?

Author: | Action                                                   | Result                                         | Next Step                               |
|----------------------------------------------------------|------------------------------------------------|-----------------------------------------|
| GET /bucket/1st-severity.json                            | Success                                        | download to a tmp directory (disk)      |
| GET /bucket/2nd-severity.zip                             | Success                                        | download to same tmp directory (disk)   |
| GET /bucket/3rd-severity.zip                             | Failed (e.g. 401, 403, 505, Timeout)           | SKIP                                    |
| Create a final zip file from all jsons in zip files      | Success                                        | Wait 4 hours for another update            |

Reviewer: Thanks, worth adding to the doc.

Can you confirm if the enricher in ClairCore will continue to serve the previous well-known score for the severity that failed?

Also, worth checking if "download to a tmp/ directory (tmp)" in Central is fine. Just be sure.

Author: > Thanks, worth adding to the doc.
> 
> Can you confirm if the enricher in ClairCore will continue to serve the previous well-known score for the severity that failed?
> 
> Also, worth checking if "download to a tmp/ directory (tmp)" in Central is fine. Just be sure.

I am pretty sure we can create tmp directory in central, I've tried a few times in other PRs but it's allowed to create regular directory. But the tmp directory eventually need to go away after we generate the single zip file so it should be ok.

Reviewer: > I am pretty sure we can create tmp directory in central, I've tried a few times in other PRs but it's allowed to create regular directory.

Thank you for checking :+1:

> so their availability should be consistently high.

This will run in customer environments, so it might not only depend on Google but on customer network setup, proxies, etc. Hence my concern on handling network issues gracefully.

@daynewlee:

> @jvdm:
>
> > Can you confirm if the enricher in ClairCore will continue to serve the previous well-known score for the severity that failed?
>
> I looked at the updateEnrichment() method in [Claircore](https://github.com/quay/claircore/blob/main/datastore/postgres/enrichment.go), seems like they intentionally block the old data after every update:

It makes sense for them to block what was updated. The scenario I am thinking of is we skip one severity (like in the example table, where we said "3rd" was skipped). But when updating the enrichments, ClairCore will mark the previous severities of the 3rd category as "outdated", effectively not showing them anymore in vulnerability matching results until we update them in the next enrichment cycle.

In simpler terms, do we need to take extra steps to avoid overwriting or blocking prior severity information when adding enrichments to ClairCore?

Reviewer 2: Curious, how did you come up with this scheme (to query by severity)?

## Consequences

> Transitioning from the NVD JSON feed to the NVD CVE API introduces certain constraints in data retrieval within the GitHub Action workflow, such as rate limiting and pagination. As a result, we will encounter a higher number of smaller, segmented JSON files.

Reviewer: > Transitioning from the NVD JSON feed to the NVD CVE API introduces certain constraints in data retrieval within the GitHub Action workflow, such as rate limiting and pagination.

Oh, I see you have some concerns about the NVD API availability. I still think it is worth clarifying the differences otherwise we can really evaluate what the constraints you're referencing are.

Changing from the NVD JSON feed to the NVD CVE API also give us the conveniences of no need to filter out data with out valid CVSS V3 metrics. 

> Following this modification, Central will no longer have a dependency on Claircore. The Claircore enricher will be used in Scanner V4, which adopts Claircore as its primary scanning engine by design.

Reviewer: > Transitioning from the NVD JSON feed to the NVD CVE API introduces certain constraints in data retrieval within the GitHub Action workflow, such as rate limiting and pagination.

:+1:

Reviewer 2: nit: ClairCore

## Context

The National Vulnerability Database (NVD) has announced the retirement of its JSON feeds by December 15, 2023. This creates the need to change our existing CVSS data updater pipeline for Scanner V4, which currently relies on these JSON feeds. 

The NVD JSON feed offers bundled data on a yearly basis, spanning from 2002 to the present year. A benefit is that only two requests are needed in the workflow for each year (one for the meta file and one for the data bundle). However, a limitation is the absence of filtering capabilities, such as by CVSS score or modification date. In contrast, the NVD API allows data retrieval through specific HTTP URL parameters, such as publish date, modified date, or CVSS score, enabling more precise data acquisition.

## Decision

The current NVD CVSS GitHub Workflow will be updated to fetch data from the NVD CVE API instead of NVD Json feeds while maintaining the freshness of the NVD CVSS data bundle in Google Storage. It will categorize the downloaded data by CVSS V3 severity levels, and these categorized data will be archived into individual zip files, such as `severity-low.zip`. The format of the JSON is the same as before.  

> We've chosen to base our API requests on CVSS V3 severity to ensure all data integrated into Claircore possesses valid CVSS V3 metrics. Consequently, retrieving data without CVSS V3 information becomes redundant.

Reviewer: ```suggestion
We've chosen to base our API requests on CVSS V3 severity to ensure all data integrated into Claircore possesses valid CVSS V3 metrics. Consequently, retrieving data without CVSS V3 information becomes irrelevant.
```

> # 0005 - Adapt Scanner V4 NVD CVSS data updating pipeline to integrate with NVD CVE Api

Reviewer 2: nit: API

- **Author(s):** Yi Li <yli3@redhat.com>
- **Created:** [2023-10-10]

## Status

Accepted

## Context

The National Vulnerability Database (NVD) has announced the retirement of its JSON feeds by December 15, 2023. This creates the need to change our existing CVSS data updater pipeline for Scanner V4, which currently relies on these JSON feeds. 

The NVD JSON feed offers bundled data on a yearly basis, spanning from 2002 to the present year. A benefit is that only two requests are needed in the workflow for each year (one for the meta file and one for the data bundle). However, a limitation is the absence of filtering capabilities, such as by CVSS score or modification date. In contrast, the NVD API allows data retrieval through specific HTTP URL parameters, such as publish date, modified date, or CVSS score, enabling more precise data acquisition.

## Decision

> The updated NVD CVSS GitHub Workflow (CI) has transitioned to sourcing data from the NVD CVE API, rather than the NVD Json feeds, ensuring the timely update of the NVD CVSS data bundle in Google Storage. To fetch data, the workflow downloads based on both the start index and CVSS v3 severity. For instance, a command such as curl "https://services.nvd.nist.gov/rest/json/cves/2.0?cvssV3Severity=CRITICAL&startIndex=0" > critical-1.json would be used. For every severity level, the workflow iteratively exhaust the startIndex to capture all associated data. And there are four severity levels in total. After all json files downloaded, they will be compress to one data bundle as one single zip file.

Reviewer 2: nit: `rather than the NVD Json feeds` -> `rather than the NVD JSON feeds`

Reviewer 2: Can you reformat the command to be more markdown friendly? For example:

To fetch data, the workflow downloads based on both the start index and CVSS v3 severity. For instance, a command such as the following:

```
curl "https://services.nvd.nist.gov/rest/json/cves/2.0?cvssV3Severity=CRITICAL&startIndex=0" > critical-1.json
```

Reviewer 2: `For every severity level, the workflow iteratively exhaust the startIndex to capture all associated data.`

What does this mean? Is this just saying we loop through calls to this endpoint via updating `startIndex`?

Also, how do we know what it is done? What indicates this? A status code? Does this content of this endpoint tell us how many vulns exist which fit the criteria?

Also, nit: `exhausts`

Reviewer 2: `And there are four severity levels in total.` What are the severity levels? Can you list them? I think this sentence isn't needed. You can just add the list of severities to the sentence "For every severity level..."

Also nit: For each severity level

Reviewer 2: nit: `After all json files downloaded, they will be compress to one data bundle as one single zip file.` -> `After all JSON files are downloaded, they will be compressed into a single ZIP file.`

Also, how did we decide on ZIP? Have we considered a tar.gz or zstd?

Reviewer 2: I took a look at how ClairCore fetches NVD data, and it assumes there is a file per year. If we use ClairCore's enricher, how will we reformat the data to fit this model? Similarly, ClairCore's enricher assumes the data is in the 1.1 JSON feed format, which is different from the 2.0 API format. I think we will need to use our own, modified version of ClairCore's CVSS enricher. This also gives us the flexibility to get more data out of NVD that ClairCore currently gets, as we may need more than what vanilla ClairCore offers

Author: "a file per year" is the json feeds that's going to be retired.
I concur with the idea of developing our own enricher. It's essential to ensure that our enricher can parse data from the Google bucket. I'll incorporate this detail into the ADR.

Reviewer 3: What is stopping us from downloading the CVEs year by year? The API appears to support this.  Or downloading all vulns (paging through the indexes) and creating the yearly files ourselves.

Author: I am not sure what's the benefit of the downloading data for every year.
We lost the conveniences of excluding data without cvss v3 and we still have the constraint of pagination. 

The yearly JSON feeds doesn't have pagination limitation so that was something nice to have. 

What do you think? @dcaravel

Author: > `For every severity level, the workflow iteratively exhaust the startIndex to capture all associated data.`
> 
> What does this mean? Is this just saying we loop through calls to this endpoint via updating `startIndex`?
> 
> Also, how do we know what it is done? What indicates this? A status code? Does this content of this endpoint tell us how many vulns exist which fit the criteria?
> 
> Also, nit: `exhausts`
In each json that's been downloaded, there are these fields to determine how many results there are in total, where the index starts and how many results per page.
resultsPerPage":2000,"startIndex":0,"totalResults":21611

Author: > nit: `After all json files downloaded, they will be compress to one data bundle as one single zip file.` -> `After all JSON files are downloaded, they will be compressed into a single ZIP file.`
> 
> Also, how did we decide on ZIP? Have we considered a tar.gz or zstd?

Zip might not be most ideal, I think we can either use `tar -xf all_json_files.tar.gz`
 or `tar -xf all_json_files.tar.xz`

> The updated NVD CVSS GitHub Workflow (CI) has transitioned to sourcing data from the NVD CVE API, rather than the NVD Json feeds, ensuring the timely update of the NVD CVSS data bundle in Google Storage. To fetch data, the workflow downloads based on both the start index and CVSS v3 severity. For instance, a command such as curl "https://services.nvd.nist.gov/rest/json/cves/2.0?cvssV3Severity=CRITICAL&startIndex=0" > critical-1.json would be used. For every severity level, the workflow iteratively exhaust the startIndex to capture all associated data. And there are four severity levels in total. After all json files downloaded, they will be compress to one data bundle as one single zip file.

Reviewer: Nit:

```suggestion
The updated NVD CVSS GitHub Workflow (CI) has transitioned to sourcing data from the NVD CVE API, rather than the NVD Json feeds, ensuring the timely update of the NVD CVSS data bundle in Google Storage. To fetch data, the workflow downloads based on both the start index and CVSS v3 severity. For instance, a command such as `curl "https://services.nvd.nist.gov/rest/json/cves/2.0?cvssV3Severity=CRITICAL&startIndex=0" > critical-1.json` would be used. For every severity level, the workflow iteratively exhaust the startIndex to capture all associated data. And there are four severity levels in total. After all json files downloaded, they will be compress to one data bundle as one single zip file.
```

> The updated NVD CVSS GitHub Workflow (CI) has transitioned to sourcing data from the NVD CVE API, rather than the NVD Json feeds, ensuring the timely update of the NVD CVSS data bundle in Google Storage. To fetch data, the workflow downloads based on both the start index and CVSS v3 severity. For instance, a command such as curl "https://services.nvd.nist.gov/rest/json/cves/2.0?cvssV3Severity=CRITICAL&startIndex=0" > critical-1.json would be used. For every severity level, the workflow iteratively exhaust the startIndex to capture all associated data. And there are four severity levels in total. After all json files downloaded, they will be compress to one data bundle as one single zip file.

Reviewer 3: The pull by severity seems OK for now to filter out entries without a CVSS v3 (assuming thats the reason why). 

This seems as though it will limit scanner's ability to detect older / newer vulnerabilities.  I understand that ClairCore currently skips anything without a CVSS v3, I assume that will NOT the case forever.  Such as when CVSS v4 becomes widespread, this pull by severity may become more complex as we will have duplicates when pulling by CVSS v3 severities and also by CVSS v4 severities.  It may make more sense to pull all the vulns (irregardless of CVSS v3 severity) and relying on either Clairs or the CI job's logic to filter out the invalid entries in a single iteration vs. dealing with duplicates in feeds.  This will also give us a datasource to address issues that come up as to 'why wasn't CVE XYZ detected', we could take a look at the CI job and see that it was filtered due to a missing score.

Author: That's a valid point. I am not sure how soon CVSS V4 score will be widely adopted. What do you guys think? @jvdm @dcaravel

Reviewer: I think when CVSS V4 scores are going to appear is not the main point, I think he's trying to ask you if grouping by severity will create any blockers in the future -- when it happens. And also, more importantly, is it just cheaper to get all vulns right now to make that future brighter and simpler? Since you're more familiar with the NVD APIs and the overall vuln updating pipeline, we would like to know if you have any thoughts about the advantages or disadvantages of pulling all vulns today or dealing with new score types later on.

Author: Even with the API key, retrieving all the data is a lengthy task. We're working to expedite the Consolidation in short term. It's hard to predict what will happen in the future and how much impact that future change would bring.
Also the question 'why wasn't CVE XYZ detected' probably won't be true, they probably would be wondering 'why CVE XYZ is missing a score'.

I think we can fetch by severity for now. 
We can always change our way of fetching data and add the filtering feature in our customized enricher (we will need it anyway to parse the json in scanner) after V4 metrics is out (or other situation happens). I feel that won't be too much of a change by that time.

We've chosen to base our API requests on CVSS V3 severity to ensure all data integrated into Claircore possesses valid CVSS V3 metrics. Consequently, retrieving data without CVSS V3 information becomes irrelevant.

> The CVSS Updater in Central will be modified to download zip file (generated by CI mentioned above) in the Google bucket. This eliminates the need for JSON parsing in the updater. ScannerV4 will pull the zip bundle from Central and populate the Matcher DB with the enrichment data.

Reviewer 2: nit: zip -> ZIP and ScannerV4 -> Scanner v4 (or Scanner V4)

> The CVSS Updater in Central will attempt to download the data bundle from the Google bucket a maximum of five times. Should all five attempts be unsuccessful, the updater will pause its efforts and initiate a fresh download attempt after a 4-hour interval. After five unsuccessful download attempts, a failure notification will be swiftly sent to the Scanner on-call Slack channel.

Reviewer 2: Maybe too much of an implementation detail for this, but will we be using some sort of exponential backoff mechanism? Will we care about the status code or any code that is not 200 nor 304 will trigger a retry?

Also, live Centrals will message our Slack channel? I don't think there is much we can do about this outside of ACSCS. I don't think we should be notified by this, as there can be many reasons downloads fail. For example, if a customer does not set up the firewall properly, there is nothing we can do, and we would not want to be spammed with errors.

We could always test out the downloads ourselves in periodic CI for sanity checks. Similar to what we have now

Author: I agree. "A failure notification will be swiftly sent to the Scanner on-call Slack channel." was too much. I will remove it.

Author: > Maybe too much of an implementation detail for this, but will we be using some sort of exponential backoff mechanism? Will we care about the status code or any code that is not 200 nor 304 will trigger a retry?
> 
> Also, live Centrals will message our Slack channel? I don't think there is much we can do about this outside of ACSCS. I don't think we should be notified by this, as there can be many reasons downloads fail. For example, if a customer does not set up the firewall properly, there is nothing we can do, and we would not want to be spammed with errors.
> 
> We could always test out the downloads ourselves in periodic CI for sanity checks. Similar to what we have now

We can use exponential backoff mechanism to retry when the error code is 429 in case there's a rate limiting. Also we can retry 5xx Server Errors, this might help with clients with proxy servers.

I agree. "A failure notification will be swiftly sent to the Scanner on-call Slack channel." was too much. I will remove it.

> The CVSS Updater in Central will attempt to download the data bundle from the Google bucket a maximum of five times. Should all five attempts be unsuccessful, the updater will pause its efforts and initiate a fresh download attempt after a 4-hour interval. After five unsuccessful download attempts, a failure notification will be swiftly sent to the Scanner on-call Slack channel.

Reviewer: 1. Promising a slack failure notification from Central is a far stretch; Central is deployed in CI, but also in Customer infra, ACSCS, and the wild, and it might be too costly to do that from those environments. Even if possible, you want to know if the bundle endpoint is healthy. We can do that by having something like a Canary (A CI job that pulls the vuln and validates its content). That's a good idea because it allows us to know if customers are experiencing issues when pulling the vuln data themselves.

2. For Central, consider creating a metric for these errors and logs. That might allow us to see those failures and why they happen if customers see the effect (stale vuln data).

3. A better retry strategy might be exponentially backing off with a large max interval (maybe hours). 
   Example:

   1. T=0s
   2. T=10s
   3. T=30s
   4. T=5min
   5. T=15min
   6. T=30min
   7. T=1h
   8. T=1h
   9. T=1h

Saying all that, for the ADR, I think it's better just to lay out these options generally, but don't get too specific unless you're sure you will implement these specifics, which I think it's hard to predict at this point. So just mentioning you are adding metrics, logs, and retrying with large intervals would be good IMO.

Author: I agree. "A failure notification will be swiftly sent to the Scanner on-call Slack channel." was too much. I will remove it.

Author: @jvdm 
My point is may be we can leave the implementation details of those in the future PR review.

Reviewer: @jvdm:

> Saying all that, for the ADR, I think it's better just to lay out these options generally, but don't get too specific unless you're sure you will implement these specifics.

:point_up: 

> I can certainly add that in Central, do we need to specify those in the doc?
> [...]
> For the exponential backoff strategy, I am assuming we don't need to specify in the doc but only to decide in implementation, is that right?

Yes. I added those values to clarify what I meant and to point out that it doesn't make sense to send a notification to Slack from Central when pulling vulns fail.

## Consequences

> One drawback of the NVD CVE API is its rate limit: only 5 requests are allowed per 30 seconds without an API Key. This restriction necessitates that our workflow fetch data at a slower pace. Additionally, transitioning from the NVD JSON feed to the NVD CVE API brings in pagination limitations. For example, when fetching data with a "low" severity, the API provides a maximum of 2000 CVEs per response. This means multiple requests are required to retrieve the complete dataset for "low" severity. The result is an increased number of smaller JSON files. Consequently, we need to compress all JSON files into a single zip archive.

Reviewer 3: What's stopping us from requesting an API key? Seems like an easy consequence to get past.

Author: We can get an API key but I just didn't want to include something undetermined in the ADR.

Changing from the NVD JSON feed to the NVD CVE API also give us the conveniences of no need to filter out data without valid CVSS V3 metrics, as mentioned above. 

Following this modification, Central will no longer have a dependency on Claircore. The Claircore enricher will be used in Scanner V4, which adopts Claircore as its primary scanning engine by design.

Per the NVD's documentation (https://nvd.nist.gov/developers/vulnerabilities), the NVD CVE API omits CVSS v3 vector strings with a 'NONE' severity. This ensures we only obtain data bearing valid CVSS V3 metrics.

> If the GH workflow (CI) encounters issues accessing or downloading portions of the CVSS data—likely due to NVD API unavailability or rate limiting—it should retry the request after a 6-second delay, capping at three total attempts. If the same HTTP request fails three times consecutively, that specific data section will be skipped. Central retains the data bundle from the prior update. In cases where the current update has data missing but the previous one doesn't (as determined by comparing the sizes of the two data bundles), the earlier data bundle will be employed to support Scanner v4.

Reviewer: That's a decision, not a consequence. Move to the "Decision" section. If you want to, mention what this decision will provide to us (in this case, stale data will be served to Scanner until the latest vulns are pulled from the bucket).

The NVD JSON feed offers bundled data on a yearly basis, spanning from 2002 to the present year. A benefit is that only two requests are needed in the workflow for each year (one for the meta file and one for the data bundle). However, a limitation is the absence of filtering capabilities, such as by CVSS score or modification date. In contrast, the NVD API allows data retrieval through specific HTTP URL parameters, such as publish date, modified date, or CVSS score, enabling more precise data acquisition.

## Decision

The updated NVD CVSS GitHub Workflow (CI) has transitioned to sourcing data from the NVD CVE API, rather than the NVD JSON feeds, ensuring the timely update of the NVD CVSS data bundle in Google Storage. The workflow is designed to download data by specifying a range of start and end dates, spanning from the year 2002 to the current year. For instance, a command listed below could be used:
```
curl "https://services.nvd.nist.gov/rest/json/cves/2.0/?pubStartDate=2021-08-04T00:00:00.000&pubEndDate=2021-10-22T00:00:00.000&startIndex=0" > 2021-test.json
``` 

> In every 120-day range, it iteratively exhausts the startIndex to capture all associated data in that date range. After all json files downloaded, they will be compressed to one data bundle as one single compressed file. We choose start and end date to be 120 days of range because this is the maximum range that NVD supports.

Reviewer 2: Instead of 120 days, I think it'd be easier to break the year down into 4 parts because it's easier to reason about it:

* January 1 - April 1
* April 1 - July 1
* July 1 - October 1
* October 1 - January 1

From here, it's still possible there are more than 2000 vulns in this range, so I'm picturing file that look like this:

`2023-A-B` where A = range number and B = index number

For example:

`2023-2-3` = vulns from 2023 from the July - October range and this is the 4th call within that range.

Does that make sense?

Author: I agree with this part. 
> January 1 - April 1
> April 1 - July 1
> July 1 - October 1
> October 1 - January 1
But maybe we can name each file as <start date> - <start index>.json
such as 2020-04-01-0.json
Then use jq to consolidate all "vunerabilities" fields to one big json file for that year to be "combine_2020.json"

Like:
```
if [ "$file_count" -gt 0 ]; then
        if jq -cs '{ vulnerabilities: map(.vulnerabilities) | add }' "$dir_name"/*"${year}"*.json > "$dir_name/combined_${year}.json"; then
            echo "The consolidated file size for $year is $(stat -c %s "$dir_name/combined_${year}.json") bytes."
```
Then remove the small sections of json data. 
So we will have 21 files in the tar file (2002 - 2023)

- **Created:** [2023-10-10]

## Status

Accepted

## Context

The National Vulnerability Database (NVD) has announced the retirement of its JSON feeds by December 15, 2023. This creates the need to change our existing CVSS data updater pipeline for Scanner V4, which currently relies on these JSON feeds. 

The NVD JSON feed offers bundled data on a yearly basis, spanning from 2002 to the present year. A benefit is that only two requests are needed in the workflow for each year (one for the meta file and one for the data bundle). However, a limitation is the absence of filtering capabilities, such as by CVSS score or modification date. In contrast, the NVD API allows data retrieval through specific HTTP URL parameters, such as publish date, modified date, or CVSS score, enabling more precise data acquisition.

## Decision

The updated NVD CVSS GitHub Workflow (CI) has transitioned to sourcing data from the NVD CVE API, rather than the NVD JSON feeds, ensuring the timely update of the NVD CVSS data bundle in Google Storage. To fetch data, the workflow downloads based on both the start index and CVSS v3 severity. For instance, a command listed below could be used: 
```

> curl "https://services.nvd.nist.gov/rest/json/cves/2.0?cvssV3Severity=CRITICAL&startIndex=0" > critical-1.json

Author: Ask product.

> curl "https://services.nvd.nist.gov/rest/json/cves/2.0?cvssV3Severity=CRITICAL&startIndex=0" > critical-1.json

Author: We need to include data with both V2 and V3 metrics. I will modify the ADR.

We've chosen to base our API requests on CVSS V3 severity to ensure all data integrated into Claircore possesses valid CVSS V3 metrics. Consequently, retrieving data without CVSS V3 information becomes irrelevant.

The CVSS Updater in Central will be modified to download compressed file (generated by CI mentioned above) in the Google bucket. This eliminates the need for JSON parsing in the updater. Scanner V4 will pull the data bundle from Central and populate the Matcher DB with the enrichment data.

A custom enricher is essential for Scanner V4, as it will provide a mechanism to retrieve JSON data from Central, parse the enrichment details, and store them in the Scanner V4 database.  Additionally, our data will be sourced from the NVD CVE API rather than JSON feeds, so the custom enricher can accommodate the distinct data formats. This also offers greater flexibility, allowing us to extract more information from the fetched JSONs than what Claircore offers.

> The CVSS Updater in Central will attempt to download the data bundle from the Google bucket using exponential retries in case of partial failures, up to a maximum interval within the next loop, which happens every 4 hours by default.

Author: Check modification date every 30 mins and only get the data when things are changed.

> If the GH workflow (CI) encounters issues accessing or downloading portions of the CVSS data—likely due to NVD API unavailability or rate limiting—it should retry the request after a 6-second delay, capping at three total attempts. If the same HTTP request fails three times consecutively, that specific data section will be skipped. Central retains the data bundle from the prior update. In cases where the current update has data missing but the previous one doesn't (as determined by comparing the sizes of the two data bundles), the earlier data bundle will be employed to support Scanner v4.

Author: If the GH workflow (CI) encounters issues accessing or downloading portions of the CVSS data—likely due to NVD API unavailability or rate limiting— the same failed http request should be retried in exponential backoff. If the all retries fail, we just abort the whole workflow and wait for next update, which is 4-hour interval.

Author: The team should be notified via Slack once the workflow fails (aborted)

Author: We need a ticket for sanity check of all bundle in GCS.

> If the GH workflow (CI) encounters issues accessing or downloading portions of the CVSS data—likely due to NVD API unavailability or rate limiting—it should retry the request after a 6-second delay, capping at three total attempts. If the same HTTP request fails three times consecutively, that specific data section will be skipped. Central retains the data bundle from the prior update. In cases where the current update has data missing but the previous one doesn't (as determined by comparing the sizes of the two data bundles), the earlier data bundle will be employed to support Scanner v4.

Author: I don't quite understand this:

> If we only skip the missing data sections, the partial bundle will still be available for Central pods that don't have the previous bundle to compare.

Why there's no previous bundle to compare? We store the bundle in blob.Snapshot from previous update

> If the GH workflow (CI) encounters issues accessing or downloading portions of the CVSS data—likely due to NVD API unavailability or rate limiting—it should retry the request after a 6-second delay, capping at three total attempts. If the same HTTP request fails three times consecutively, that specific data section will be skipped. Central retains the data bundle from the prior update. In cases where the current update has data missing but the previous one doesn't (as determined by comparing the sizes of the two data bundles), the earlier data bundle will be employed to support Scanner v4.

Reviewer: In any case, my question remains:

> Are we skipping the entire bundle or just the missing data sections?

When you say, "the previous bundle will be employed", I understood you will serve the previous bundle whole to Scanner, and ignore the new one with missing data sections. I just wanted to clarify that.

> If the GH workflow (CI) encounters issues accessing or downloading portions of the CVSS data—likely due to NVD API unavailability or rate limiting—it should retry the request after a 6-second delay, capping at three total attempts. If the same HTTP request fails three times consecutively, that specific data section will be skipped. Central retains the data bundle from the prior update. In cases where the current update has data missing but the previous one doesn't (as determined by comparing the sizes of the two data bundles), the earlier data bundle will be employed to support Scanner v4.

Author: Yes, the bundle is stored in file system. If the data is incomplete and the pod restarts then there will be no previous bundle to compare. 

And we will just skip the missing data sections. 
> To clarify, are we skipping the entire bundle or just the missing data sections?

## Consequences

One drawback of the NVD CVE API is its rate limit: only 5 requests are allowed per 30 seconds without an API Key. This restriction necessitates that our workflow fetch data at a slower pace. Additionally, transitioning from the NVD JSON feed to the NVD CVE API brings in pagination limitations. For example, when fetching data with a "low" severity, the API provides a maximum of 2000 CVEs per response. This means multiple requests are required to retrieve the complete dataset for "low" severity. The result is an increased number of smaller JSON files. Consequently, we need to compress all JSON files into a single archived file such as tar.gz or tar.xz.

Changing from the NVD JSON feed to the NVD CVE API also give us the conveniences of no need to filter out data without valid CVSS V3 metrics, as mentioned above. 

Following this modification, Central will no longer have a dependency on Claircore. The Claircore enricher will be used in Scanner V4, which adopts Claircore as its primary scanning engine by design.

> 

Author: Since we will abort the workflow once the data is incomplete, we might not be able to get most updated bundle in quite a few hours. But if the workflow is aborted, the team will get notified.

The CVSS Updater in Central will be modified to download compressed file (generated by CI mentioned above) in the Google bucket. This eliminates the need for JSON parsing in the updater. Scanner V4 will pull the data bundle from Central and populate the Matcher DB with the enrichment data.

A custom enricher is essential for Scanner V4, as it will provide a mechanism to retrieve JSON data from Central, parse the enrichment details, and store them in the Scanner V4 database.  Additionally, our data will be sourced from the NVD CVE API rather than JSON feeds, so the custom enricher can accommodate the distinct data formats. This also offers greater flexibility, allowing us to extract more information from the fetched JSONs than what Claircore offers.

The CVSS Updater in Central will attempt to download the data bundle from the Google bucket using exponential retries in case of partial failures, up to a maximum interval within the next loop, which happens every 4 hours by default.

> Should the GitHub Actions workflow (CI) experience difficulties in accessing or downloading segments of the CVSS data due to an HTTP request failure—potentially caused by the NVD API being unavailable or due to rate limiting—the affected HTTP request should be automatically retried following an exponential backoff strategy. In the event that all retry attempts fail, the workflow will be terminated, and the system will await the next scheduled update in 4 hours. The team will be promptly notified of any aborted or failed workflows via Slack.

Reviewer 2: It is currently 3 hours in CI. Are we making it 4 hours now?

Also, can we have a way to trigger it manually in case we detect it's up and we don't want to wait?

Author: We can keep it as 3 hours.
As long as we use `workflow_dispatch` in workflow, we can manually trigger a workflow.

## Consequences

> One drawback of the NVD CVE API is its rate limit: only 5 requests are allowed per 30 seconds without an API Key. This restriction necessitates that our workflow fetch data at a slower pace. Additionally, transitioning from the NVD JSON feed to the NVD CVE API brings in pagination limitations. For example, when fetching data with a "low" severity, the API provides a maximum of 2000 CVEs per response. This means multiple requests are required to retrieve the complete dataset for "low" severity. The result is an increased number of smaller JSON files. Consequently, we need to compress all JSON files into a single archived file such as tar.gz or tar.xz.

Reviewer 2: Another option is to buffer the vulnerabilities until we hit maybe 20,000 (random number) and then write them to a file. Have we considered this?

Reviewer: Buffering in memory? Any benefit to the producer or consumer compared to using files?

Author: We probably don't need to decompress that tar file so we should be good.

Reviewer: I expect we will decompress the tar file when parsing the JSON to insert it into the DB. My question was, what is the benefit of buffering in memory to create larger files vs. using smaller files?

Author: I discussed with @RTann . He asked the question because he was concerned there will be many json files to be decompressed in scanner. But for the gzreader we can actually iterate the files in tar. @jvdm

# 0005 - Adapt Scanner V4 NVD CVSS data updating pipeline to integrate with NVD CVE Api

- **Author(s):** Yi Li <yli3@redhat.com>
- **Created:** [2023-10-10]

## Status

Accepted

## Context

The National Vulnerability Database (NVD) has announced the retirement of its JSON feeds by December 15, 2023. This creates the need to change our existing CVSS data updater pipeline for Scanner V4, which currently relies on these JSON feeds. 

The NVD JSON feed offers bundled data on a yearly basis, spanning from 2002 to the present year. A benefit is that only two requests are needed in the workflow for each year (one for the meta file and one for the data bundle). However, a limitation is the absence of filtering capabilities, such as by CVSS score or modification date. In contrast, the NVD API allows data retrieval through specific HTTP URL parameters, such as publish date, modified date, or CVSS score, enabling more precise data acquisition.

## Decision

> The current NVD CVSS GitHub Workflow will be updated to fetch data from the NVD CVE API instead of NVD Json feeds while maintaining the freshness of the NVD CVSS data bundle in Google Storage. It will categorize the downloaded data by CVSS V3 severity levels, and these categorized data will be archived into individual zip files, such as `severity-low.zip`. The format of the JSON is the same as before.  

Reviewer 2: Can you give an example query with the associated sample zip file its contents would go into?

> The current NVD CVSS GitHub Workflow will be updated to fetch data from the NVD CVE API instead of NVD Json feeds while maintaining the freshness of the NVD CVSS data bundle in Google Storage. It will categorize the downloaded data by CVSS V3 severity levels, and these categorized data will be archived into individual zip files, such as `severity-low.zip`. The format of the JSON is the same as before.  

Author: curl "https://services.nvd.nist.gov/rest/json/cves/2.0?cvssV3Severity=CRITICAL&startIndex=0&endIndex=5" > critical-2001.json

In reality we don't need the **endIndex**

> We've chosen to base our API requests on CVSS V3 severity to ensure all data integrated into Claircore possesses valid CVSS V3 metrics. Consequently, retrieving data without CVSS V3 information becomes irrelevant.

Author: Claircore only cares about data with valid CVSS v3 score. https://github.com/quay/claircore/blob/main/enricher/cvss/feed.go Line 58-69 

And we are only retrieving data with CVSS v3 score. Does that answer your question?

The CVSS Updater in Central will be modified to download zip files corresponding to four distinct severity levels and write them into a single zip file stored in Central's file system. This eliminates the need for JSON parsing in the updater. ScannerV4 will pull the zip bundle and populate the Matcher DB with the enrichment data.

> The CVSS Updater will attempt to download each zip file from the Google bucket up to five times. If all five attempts fail, it will skip that particular file. For instance

Reviewer 2: Is this referring to Central updater or CI?

Author: This is referring to central updater. For updater in CI, I used "workflow" in the document.

| Action                                                   | Result                                         | Next Step                               |
|----------------------------------------------------------|------------------------------------------------|-----------------------------------------|
| GET /bucket/1st-severity.zip                            | Success                                        | download to a tmp directory (disk)      |
| GET /bucket/2nd-severity.zip                             | Success                                        | download to same tmp directory (disk)   |
| GET /bucket/3rd-severity.zip                             | Failed (e.g. 401, 403, 505, Timeout)           | Retry 5 times and SKIP                  |

> | Create a final zip file from all jsons in downloaded zip files and save to store.Snapshot      | Success                                        | Wait 4 hours for another update         |

Reviewer 2: we used to do 3 hours. We changed to 4 due to OpenShift CI's... efficiencies...

How often does Clair fetch vuln updates? Do we know about how often NVD's data may change (not sure if it's daily, hourly, weekly, etc)?

Author: https://nvd.nist.gov/vuln/data-feeds says:
The APIs are updated as frequently as our website (unlike the traditional feeds which have explicit update intervals)

https://nvd.nist.gov/general/FAQ-Sections/General-FAQs#faqLink8 also says:
Q. How often is the data in the NVD updated?
 A. CVEs are typically available in the NVD within an hour of their publishing. Once a CVE is in the NVD, analysts can begin the analysis process. The processing time can vary depending on the CVE, the information available, and the quantity of CVEs published within a given time frame. The NVD updates CVE records as soon as analysis is complete. CVE data and/or CPE names may later be updated . If modification occurs, the NVD will automatically refresh any relevant information.

> ## Consequences

Reviewer 2: do we know about how large each ZIP would be at the moment? Have we considered other compression formats? For example, we talked about `zstd` for vuln data. Would that be best here, too?

One drawback of the NVD CVE API is its rate limit: only 5 requests are allowed per 30 seconds without an API Key. This restriction necessitates that our workflow fetch data at a slower pace. Additionally, transitioning from the NVD JSON feed to the NVD CVE API brings in pagination limitations. For example, when fetching data with a "low" severity, the API provides a maximum of 2000 CVEs per response. This means multiple requests are required to retrieve the complete dataset for "low" severity. The result is an increased number of smaller JSON files. Consequently, we need to compress all JSON files for a particular severity into a single zip archive.

Changing from the NVD JSON feed to the NVD CVE API also give us the conveniences of no need to filter out data with out valid CVSS V3 metrics, as mentioned above. 

Following this modification, Central will no longer have a dependency on Claircore. The Claircore enricher will be used in Scanner V4, which adopts Claircore as its primary scanning engine by design.

> Per the NVD's documentation (https://nvd.nist.gov/General/News/change-timeline), the NVD CVE API omits CVSS v3 vector strings with a 'NONE' severity. This ensures we only obtain data bearing valid CVSS V3 metrics.

Reviewer 2: nit: `Per [NVD's documentation](https://nvd.nist.gov/General/News/change-timeline)`

Reviewer 2: I don't think I see this information on this page. Is this the right link?

> If there are discrepancies in the CVSS data during fetching or downloading, the index report will show an incomplete CVSS score, and certain CVEs might be without their respective scores. Because UpdateEnrichments in Claircore initiates a new UpdateOperation, adds the provided EnrichmentRecord(s), and ensures that enrichments from past updates are not accessible to clients.

Author: Do you have any recommendations regarding this matter? Is it acceptable for customers to have missing CVSS scores in their scanning reports for a few hours? @RTann 

cc: @jvdm

Reviewer 2: I think it is very possible we will be missing CVSS scores for vulnerabilities, as it may take NVD time to add the vulnerability into its database. This is something we deal with now, but we opt to block those vulns until CVSS scores are present (or we manually add the CVEs ourselves).

I think it's best to allow this scenario, and call CVEs without CVSS scores valid. No reason a user should not know about a potential vulnerability just because it's missing a CVSS score. The vendor may have assigned a severity which should help the user assess the risk

Reviewer: @ross The scenario Yi is calling out is when we cannot pull CVSS scores for a given "group" (in his proposal, severities). He's proposing we attempt to update the other successful severities, which means we will effectively delete the missing ones since ClairCore handles these updates as if they were "Snapshots" rather than insertions/append.

Reviewer: ```suggestion
When some parts of the CVSS data cannot be accessed or downloaded during the process of fetching or downloading, the vulnerability report may show missing CVSS scores for some vulns that had them before. This happens because the `UpdateEnrichments` feature in ClairCore adds the EnrichmentRecord(s) provided, but it ensures that the enrichments from previous updates cannot be accessed by clients, effectively removing them from vulnerability reports.```

Reviewer 4: You've got the wrong @ross

## Decision

The current NVD CVSS GitHub Workflow will be updated to fetch data from the NVD CVE API instead of NVD Json feeds while maintaining the freshness of the NVD CVSS data bundle in Google Storage. It will categorize the downloaded data by CVSS V3 severity levels, and these categorized data will be archived into individual zip files, such as `severity-low.zip`. The format of the JSON is the same as before.  

We've chosen to base our API requests on CVSS V3 severity to ensure all data integrated into Claircore possesses valid CVSS V3 metrics. Consequently, retrieving data without CVSS V3 information becomes redundant.

The CVSS Updater in Central will be modified to download zip files corresponding to four distinct severity levels and write them into a single zip file stored in Central's file system. This eliminates the need for JSON parsing in the updater. ScannerV4 will pull the zip bundle and populate the Matcher DB with the enrichment data.

## Consequences

> One drawback of the NVD CVE API is its rate limit: only 5 requests are allowed per 30 seconds without an API Key. This restriction necessitates that our workflow fetch data at a slower pace. Additionally, transitioning from the NVD JSON feed to the NVD CVE API brings in pagination limitations. For example, when fetching data with a "low" severity, the API provides a maximum of 2000 CVEs per response. This means multiple requests are required to retrieve the complete dataset for "low" severity. The result is an increased number of smaller JSON files. Consequently, we need to compress all JSON files for a particular severity into a single zip archive.

Reviewer: > This restriction necessitates that our workflow fetch data at a slower pace

I [made one page
request](https://services.nvd.nist.gov/rest/json/cves/2.0?cvssV3Severity=CRITICAL&startIndex=0),
and the results say:

```
{"resultsPerPage":2000,"startIndex":0,"totalResults":21470...
```

That's roughly ten pages per severity. It would take ~1 minute to get all pages at ~0.15 pages per second (1 request / 6s). Fine.

It might be rate limiting is more nuanced than this, but I think it's fine to go for it and iterate.

Reviewer 2: I think it may be worth asking if we can get an API key so we aren't as limited: https://nvd.nist.gov/developers/request-an-api-key

^ before filling that out we'd definitely need to ask for permission

Reviewer: Or that, although I don't think it blocks the ADR at this point.
